File: src/services/auth-service.ts
export type AuthState = "authenticated" | "missing_token";

export interface AuthStatus {
  state: AuthState;
  accountDisplayName?: string;
  lastAuthenticatedAt?: string;
}

export interface AuthStartResult {
  authorizeUrl: string;
  state: string;
  callbackUrl: string;
}

export interface AuthService {
  startAuth(): Promise<AuthStartResult>;
  getStatus(): Promise<AuthStatus>;
  logout(): Promise<AuthStatus>;
}

export class MockAuthService implements AuthService {
  private authenticated = false;

  async startAuth(): Promise<AuthStartResult> {
    const state = "mock-state-123";
    this.authenticated = true;
    return {
      authorizeUrl: "https://www.bungie.net/en/oauth/authorize?client_id=mock",
      state,
      callbackUrl: "http://127.0.0.1:8787/oauth/callback"
    };
  }

  async getStatus(): Promise<AuthStatus> {
    if (!this.authenticated) {
      return { state: "missing_token" };
    }

    return {
      state: "authenticated",
      accountDisplayName: "MockGuardian#1234",
      lastAuthenticatedAt: new Date().toISOString()
    };
  }

  async logout(): Promise<AuthStatus> {
    this.authenticated = false;
    return { state: "missing_token" };
  }
}

File: src/vaultpilot/services/auth_service.py
"""Auth service protocol + a mock implementation.

The real implementation (M2) will drive the Bungie OAuth2 authorization-code
flow through a localhost callback and persist tokens. The mock just toggles an
in-memory authenticated flag so the auth tools have predictable behavior.
"""

from __future__ import annotations

from typing import Protocol, TypedDict


class AuthStartResult(TypedDict):
    authorize_url: str
    state: str
    callback_url: str


class AuthStatus(TypedDict, total=False):
    state: str  # "authenticated" | "missing_token"
    account_display_name: str


class AuthService(Protocol):
    def start_auth(self) -> AuthStartResult: ...
    def get_status(self) -> AuthStatus: ...
    def logout(self) -> AuthStatus: ...


class MockAuthService:
    """In-memory mock. No real OAuth, no network, no tokens."""

    def __init__(self) -> None:
        self._authenticated = False

    def start_auth(self) -> AuthStartResult:
        # Real flow will generate a CSRF state nonce and a real authorize URL.
        self._authenticated = True
        return {
            "authorize_url": "https://www.bungie.net/en/OAuth/Authorize?client_id=mock&response_type=code",
            "state": "mock-state-123",
            "callback_url": "https://localhost:7777/callback",
        }

    def get_status(self) -> AuthStatus:
        if not self._authenticated:
            return {"state": "missing_token"}
        return {"state": "authenticated", "account_display_name": "MockGuardian#1234"}

    def logout(self) -> AuthStatus:
        self._authenticated = False
        return {"state": "missing_token"}
